src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,334 @@
+                foreach (var documentSymbol in documentSymbolData)
+                {
+                    var sortedChildren = SortDocumentSymbols(documentSymbol.Children, sortOption, cancellationToken);
+                    sortedDocumentSymbols.Add(new DocumentSymbolData(documentSymbol, sortedChildren));
+                }
+
+                switch (sortOption)
+                {
+                    case SortOption.Name:
+                        sortedDocumentSymbols.Sort(static (x, y) => x.Name.CompareTo(y.Name));
+                        break;
+                    case SortOption.Location:
+                        sortedDocumentSymbols.Sort(static (x, y) => x.RangeSpan.Start - y.RangeSpan.Start);
+                        break;
+                    case SortOption.Type:
+                        sortedDocumentSymbols.Sort(static (x, y) =>
+                        {
+                            if (x.SymbolKind == y.SymbolKind)
+                                return x.Name.CompareTo(y.Name);
+
+                            return x.SymbolKind - y.SymbolKind;
+                        });
+                        break;
+                    default:
+                        ExceptionUtilities.UnexpectedValue(sortOption);
+                        break;
+                }
+
+                return sortedDocumentSymbols.ToImmutable();
+            }
+        }
+
+        /// <summary>
+        /// Returns an immutable array of DocumentSymbolData such that each node matches the given pattern.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolData> SearchDocumentSymbolData(
+            ImmutableArray<DocumentSymbolData> documentSymbolData,
+            string pattern,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var _ = ArrayBuilder<DocumentSymbolData>.GetInstance(out var filteredDocumentSymbols);
+            var patternMatcher = PatternMatcher.CreatePatternMatcher(pattern, includeMatchedSpans: false, allowFuzzyMatching: true);
+
+            foreach (var documentSymbol in documentSymbolData)
+            {
+                var filteredChildren = SearchDocumentSymbolData(documentSymbol.Children, pattern, cancellationToken);
+                if (SearchNodeTree(documentSymbol, patternMatcher, cancellationToken))
+                    filteredDocumentSymbols.Add(new DocumentSymbolData(documentSymbol, filteredChildren));
+            }
+
+            return filteredDocumentSymbols.ToImmutable();
+
+            // Returns true if the name of one of the tree nodes results in a pattern match.
+            static bool SearchNodeTree(DocumentSymbolData tree, PatternMatcher patternMatcher, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (patternMatcher.Matches(tree.Name))
```

Comment:
what would be slick in the future is if we captured the text-spans of the match so we could *highlight* it in the tree.  e.g. if the symbol name is "WriteLine" and the user writes "WL" in teh search box, then we would highlight the "W" and "L" in "WriteLine".  If you have a tracking list of future ideas, can you add this here?